File: frontend/src/lib/workspaceMedia.ts
export type ChatMediaAttachment = {
  path: string
  kind: 'image' | 'animation' | 'video'
  label?: string
}

const MEDIA_EXT =
  /\.(png|jpe?g|gif|webp|svg|webm|mp4|apng)(?:\?.*)?$/i

const MARKDOWN_IMAGE_RE = /!\[[^\]]*\]\(([^)]+)\)/g
const MEDIA_PATH_RE =
  /(?:^|[\s`"'(])([^\s`"')]+\.(?:png|jpe?g|gif|webp|svg|webm|mp4|apng))/gi

function mediaKind(path: string): ChatMediaAttachment['kind'] {
  const ext = path.split('.').pop()?.toLowerCase() || ''
  if (ext === 'gif' || ext === 'webp' || ext === 'apng') return 'animation'
  if (ext === 'webm' || ext === 'mp4') return 'video'
  return 'image'
}

export function workspaceMediaUrl(path: string): string {
  const normalized = path.replace(/^\/+/, '')
  return `/api/workspace/media/${encodeURI(normalized)}`
}

export function isWorkspaceMediaPath(value: string): boolean {
  const text = value.trim()
  if (!text || /^https?:\/\//i.test(text) || text.startsWith('data:')) return false
  return MEDIA_EXT.test(text)
}

export function normalizeWorkspaceMediaPath(raw: string): string | null {
  const text = raw.trim().replace(/^['"]|['"]$/g, '')
  if (!isWorkspaceMediaPath(text)) return null

  const repoMarker = '/backtrading-demo/'
  const idx = text.indexOf(repoMarker)
  if (idx >= 0) {
    return text.slice(idx + repoMarker.length)
  }

  if (text.startsWith('/')) {
    const parts = text.split('/').filter(Boolean)
    return parts.slice(-2).join('/') || parts[parts.length - 1] || null
  }

  return text.replace(/^\.\//, '')
}

export function extractMediaPathsFromText(text: string): string[] {
  if (!text) return []

  const found: string[] = []
  const seen = new Set<string>()

  const add = (raw: string) => {
    const path = normalizeWorkspaceMediaPath(raw)
    if (!path || seen.has(path)) return
    seen.add(path)
    found.push(path)
  }

  for (const match of text.matchAll(MARKDOWN_IMAGE_RE)) {
    add(match[1])
  }
  for (const match of text.matchAll(MEDIA_PATH_RE)) {
    add(match[1])
  }

  return found
}

export function attachmentsFromPaths(paths: string[]): ChatMediaAttachment[] {
  const rows: ChatMediaAttachment[] = []
  const seen = new Set<string>()

  for (const raw of paths) {
    const path = normalizeWorkspaceMediaPath(raw)
    if (!path || seen.has(path)) continue
    seen.add(path)
    rows.push({
      path,
      kind: mediaKind(path),
      label: path.split('/').pop(),
    })
  }

  return rows
}

export function extractMediaAttachments(text: string): ChatMediaAttachment[] {
  return attachmentsFromPaths(extractMediaPathsFromText(text))
}

export function mergeAttachments(
  ...groups: Array<ChatMediaAttachment[] | undefined>
): ChatMediaAttachment[] {
  const rows: ChatMediaAttachment[] = []
  const seen = new Set<string>()

  for (const group of groups) {
    for (const item of group || []) {
      if (!item?.path || seen.has(item.path)) continue
      seen.add(item.path)
      rows.push({
        path: item.path,
        kind: item.kind || mediaKind(item.path),
        label: item.label || item.path.split('/').pop(),
      })
    }
  }

  return rows
}

export function resolveMarkdownImageSrc(src: string | undefined): string | undefined {
  if (!src) return undefined
  const trimmed = src.trim()
  if (/^https?:\/\//i.test(trimmed) || trimmed.startsWith('data:')) return trimmed
  const path = normalizeWorkspaceMediaPath(trimmed)
  return path ? workspaceMediaUrl(path) : trimmed
}

File: api/workspace_media.py
"""Serve workspace image/chart/animation files to the UI with path traversal protection."""

from __future__ import annotations

import mimetypes
import re
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse

from control_plane.engine_process_manager import REPO_ROOT

router = APIRouter(prefix="/api/workspace", tags=["workspace-media"])

MEDIA_EXTENSIONS = frozenset({
    ".png",
    ".jpg",
    ".jpeg",
    ".gif",
    ".webp",
    ".svg",
    ".webm",
    ".mp4",
    ".apng",
})

MARKDOWN_IMAGE_RE = re.compile(r"!\[[^\]]*\]\(([^)]+)\)")
MEDIA_PATH_RE = re.compile(
    r"(?:^|[\s`\"'(])([^\s`\"')]+\.(?:png|jpe?g|gif|webp|svg|webm|mp4|apng))",
    re.IGNORECASE,
)


def media_kind_for_suffix(suffix: str) -> str:
    lowered = suffix.lower()
    if lowered in {".gif", ".webp", ".apng"}:
        return "animation"
    if lowered in {".webm", ".mp4"}:
        return "video"
    return "image"


def resolve_workspace_media_path(raw: str) -> Path | None:
    """Resolve a repo-relative or absolute-in-repo path to a readable file."""
    text = (raw or "").strip().strip("'\"")
    if not text or text.startswith(("http://", "https://", "data:")):
        return None

    candidate = Path(text)
    if not candidate.is_absolute():
        candidate = (REPO_ROOT / candidate).resolve()
    else:
        candidate = candidate.resolve()

    try:
        candidate.relative_to(REPO_ROOT.resolve())
    except ValueError:
        return None

    if not candidate.is_file():
        return None
    if candidate.suffix.lower() not in MEDIA_EXTENSIONS:
        return None
    return candidate


def extract_media_paths_from_text(text: str) -> list[str]:
    """Find workspace media paths referenced in markdown or plain text."""
    if not text:
        return []

    found: list[str] = []
    seen: set[str] = set()

    def add(raw: str) -> None:
        cleaned = raw.strip().strip("'\"")
        if not cleaned or cleaned in seen:
            return
        resolved = resolve_workspace_media_path(cleaned)
        if resolved is None:
            return
        try:
            rel = resolved.relative_to(REPO_ROOT.resolve()).as_posix()
        except ValueError:
            return
        seen.add(cleaned)
        seen.add(rel)
        found.append(rel)

    for match in MARKDOWN_IMAGE_RE.finditer(text):
        add(match.group(1))

    for match in MEDIA_PATH_RE.finditer(text):
        add(match.group(1))

    return found


def attachments_from_paths(paths: list[str]) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    seen: set[str] = set()
    for raw in paths:
        resolved = resolve_workspace_media_path(raw)
        if resolved is None:
            continue
        rel = resolved.relative_to(REPO_ROOT.resolve()).as_posix()
        if rel in seen:
            continue
        seen.add(rel)
        rows.append({
            "path": rel,
            "kind": media_kind_for_suffix(resolved.suffix),
            "label": resolved.name,
        })
    return rows


def extract_media_attachments(text: str) -> list[dict[str, Any]]:
    return attachments_from_paths(extract_media_paths_from_text(text))


@router.get("/media/{file_path:path}")
async def get_workspace_media(file_path: str):
    resolved = resolve_workspace_media_path(file_path)
    if resolved is None:
        raise HTTPException(status_code=404, detail="Media file not found")
    media_type, _ = mimetypes.guess_type(resolved.name)
    return FileResponse(
        resolved,
        media_type=media_type or "application/octet-stream",
        filename=resolved.name,
    )
